internal/providers/dockerhub/dockerhub.go

```diff
@@ -17,51 +17,87 @@ package dockerhub
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
 
-	"github.com/distribution/reference"
-	"github.com/docker/docker/api/types/registry"
-	"github.com/moby/moby/distribution"
+	"golang.org/x/oauth2"
 
 	provifv1 "github.com/stacklok/minder/pkg/providers/v1"
 )
 
 // DockerHub is the struct that contains the Docker Hub client
 type DockerHub struct {
+	cred      provifv1.OAuth2TokenCredential
+	cli       *http.Client
 	namespace string
+	target    *url.URL
 }
 
-// Ensure that the Docker Hub client implements the DockerHub interface
-var _ provifv1.DockerHub = (*DockerHub)(nil)
+// Ensure that the Docker Hub client implements the ContainerLister interface
+var _ provifv1.ContainerLister = (*DockerHub)(nil)
 
 // New creates a new Docker Hub client
-func New(ns string) (*DockerHub, error) {
+func New(cred provifv1.OAuth2TokenCredential, ns string) (*DockerHub, error) {
+	cli := oauth2.NewClient(context.Background(), cred.GetAsOAuth2TokenSource())
+
+	u, err := url.Parse("https://hub.docker.com/v2/repositories")
+	if err != nil {
+		return nil, fmt.Errorf("error parsing base URL: %w", err)
+	}
+
+	t := u.JoinPath(ns)
+
 	return &DockerHub{
 		namespace: ns,
+		cred:      cred,
+		cli:       cli,
+		target:    t,
 	}, nil
 }
 
 // ListContainers lists the containers in the Docker Hub
 func (d *DockerHub) ListContainers(ctx context.Context) ([]string, error) {
-	ref, err := reference.ParseNamed(d.namespace)
+	req, err := http.NewRequest("GET", d.target.String(), nil)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating request: %w", err)
 	}
 
-	repos, err := distribution.GetRepositories(ctx, ref, &distribution.ImagePullConfig{
-		Config: distribution.Config{
-			AuthConfig: &registry.AuthConfig{
-				// TODO add auth
-			},
-		},
-	})
+	resp, err := d.cli.Do(req.WithContext(ctx))
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error making request: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusUnauthorized {
+			return nil, fmt.Errorf("unauthorized: %s", resp.Status)
```

Comment:
These errors look like something we'll want to define our error types for so we can later do errors.Is in the callers

Reply:
trve